### PR TITLE
Add typeroots support

### PR DESCRIPTION
**Pull request by tomv564**
On codebases that rely on type roots for distributing extra declaration files, this code should prevent diagnostics about missing types and improve completions/hovers/etc.

To do: 
- [x] Fix lint error
- [x] Add test(s)

### CODE REVIEW

**Comment by tomv564** [comment]
@felixfbecker, I am a bit stuck on this test - the types.d.ts file is added to the in-memory file system but its content is not loaded. What am I missing?

**Comment by tomv564** [comment]
@felixfbecker The content-preload for declaration files was missing typeRoots, tests pass with windows and unix-style paths now.
I'm not happy with having introduced yet another `ensure` function, but couldn't find any other place that loads content based on tsconfig settings?

File: src/project-manager.ts [hunk 1]
```diff
@@ -337,6 +338,24 @@ export class ProjectManager implements Disposable {
...
+		}
+		return false;
+	}
+
+	ensureConfigDependencies(): Observable<never> {
+		return observableFromIterable(this.inMemoryFs.uris())
+			.filter(uri => this.isConfigDependency(uri2path(uri)))
+			.mergeMap(uri => this.updater.ensure(uri))
+			.publishReplay()
+			.refCount();
```
Comment (felixfbecker) [inline src/project-manager.ts]: why publishReplay + refCount?

File: src/project-manager.ts [hunk 2]
```diff
@@ -838,6 +862,11 @@ export class ProjectConfiguration {
 		this.expectedFilePaths = new Set(configParseResult.fileNames);
 
 		const options = configParseResult.options;
+		const pathResolver = /[a-z]:\//i.test(base) ? path.win32 : path.posix;
```
Comment (felixfbecker) [inline src/project-manager.ts]: the regexp should only match at the beginning of the string

File: src/project-manager.ts [hunk 3]
```diff
@@ -865,6 +894,19 @@ export class ProjectConfiguration {
 	private ensuredBasicFiles = false;
 
 	/**
+	 * [isExpectedDeclarationFile description]
+	 * @param {string} fileName [description]
```
Comment (felixfbecker) [inline src/project-manager.ts]: template?

File: src/project-manager.ts (hunk 1 above)
Comment (tomv564) [reply src/project-manager.ts]: I looked at ensureModuleStructure, which I also understood would be invoked many times but should always return the same (cached) sequence. But I'd love to know how this use case differs (and remove it if unnecessary).

**Comment by tomv564** [comment]
@felixfbecker: would you have time to look at this PR again soon?

File: src/project-manager.ts (hunk 1 above)
Comment (felixfbecker) [reply src/project-manager.ts]: this doesn't actually save the (i.e. cache) the Observable, so it doesn't has a use here. If you want to always return the same Observable you need to save it in a property

File: src/project-manager.ts [hunk 4]
```diff
@@ -735,6 +754,11 @@ export class ProjectConfiguration {
 	private expectedFilePaths = new Set<string>();
 
 	/**
+	 * List of resolved extra root directories to allow global type declaration files to be loaded from.
```
Comment (felixfbecker) [inline src/project-manager.ts]: Is this restricted to global typings?

File: src/project-manager.ts [hunk 5]
```diff
@@ -735,6 +754,11 @@ export class ProjectConfiguration {
...
+	 */
+	private typeRoots: string[];
```
Comment (felixfbecker) [inline src/project-manager.ts]: is the order important? if not, use a Set

File: src/project-manager.ts [hunk 6]
```diff
@@ -865,6 +894,19 @@ export class ProjectConfiguration {
...
+	 * Determines if a fileName is a declaration file within expected files or type roots
+	 * @param fileName
+	 */
+	public isExpectedDeclarationFile(fileName: string) {
+		if (isDeclarationFile(fileName)) {
+			return this.expectedFilePaths.has(toUnixPath(fileName)) ||
+					this.typeRoots.some(root => fileName.startsWith(root));
+		} else {
+			return false;
+		}
```
Comment (felixfbecker) [inline src/project-manager.ts]: this function just needs to be a single boolean `return`

File: src/project-manager.ts [hunk 7]
```diff
@@ -865,6 +894,19 @@ export class ProjectConfiguration {
 	private ensuredBasicFiles = false;
 
 	/**
+	 * Determines if a fileName is a declaration file within expected files or type roots
+	 * @param fileName
+	 */
+	public isExpectedDeclarationFile(fileName: string) {
```
Comment (felixfbecker) [inline src/project-manager.ts]: return type

File: src/project-manager.ts [hunk 8]
```diff
@@ -337,6 +338,24 @@ export class ProjectManager implements Disposable {
 		});
 	}
 
+	isConfigDependency(filePath: string): boolean {
```
Comment (felixfbecker) [inline src/project-manager.ts]: please add a docblock

File: src/project-manager.ts [hunk 9]
```diff
@@ -337,6 +338,24 @@ export class ProjectManager implements Disposable {
...
+		for (const config of this.configurations()) {
+			config.ensureConfigFile();
+			if (config.isExpectedDeclarationFile(filePath)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	ensureConfigDependencies(): Observable<never> {
```
Comment (felixfbecker) [inline src/project-manager.ts]: please add a docblock

File: src/project-manager.ts (hunk 4 above)
Comment (tomv564) [reply src/project-manager.ts]: From https://www.typescriptlang.org/docs/handbook/tsconfig-json.html:
> By default all visible “@types” packages are included in your compilation.
> If typesRoots is specified, only packages under typeRoots will be included.
> Keep in mind that automatic inclusion is only important if you’re using files with global declarations (as opposed to files declared as modules)

File: src/project-manager.ts (hunk 5 above)
Comment (tomv564) [reply src/project-manager.ts]: Not sure if `this.typeRoots.some(root => fileName.startsWith(root));` is a good fit for Set?

File: src/project-manager.ts (hunk 4 above)
Comment (tomv564) [reply src/project-manager.ts]: They clarify in that page as well: 
> If you use an import "foo" statement, for instance, TypeScript may still look through node_modules & node_modules/@types folders to find the foo package.

**Comment by tomv564** [comment]
@felixfbecker: your suggestions have been applied (with the exception of Set, see comment)

**Comment by tomv564** [comment]
Ping @felixfbecker, congrats on the big release today. Any chance we can work towards merging this PR and the plugin support PR?

**Comment by felixfbecker** [comment]
Thank you, as you can imagine I was very busy in the last days :) Taking a look

File: src/project-manager.ts [hunk 10]
```diff
@@ -321,6 +321,7 @@ export class ProjectManager implements Disposable {
 			span.addTags({ uri, maxDepth });
 			ignore.add(uri);
 			return this.ensureModuleStructure(span)
+				.concat(Observable.defer(() => this.ensureConfigDependencies()))
```
Comment (felixfbecker) [inline src/project-manager.ts]: Does this have a performance impact? I think it probably has, as you are iterating in `ensureConfigDependencies()` over every single file in the workspace, and `ensureReferencedFiles()` is called recursively for every referenced file, so you are repeating that iteration every time. I think `ensureConfigDependencies()` needs to memoize its return value in the same way `ensureAllFiles()` does, or you could just add that logic to `ensureModuleStructure()` (which would already is memoized and also has memoize cache invalidation logic).

**Comment by tomv564** [comment]
@felixfbecker I applied your suggestion, and verified that:
* The logic still works with a `typeRoots` dependent project
* Hovers on the Microsoft/Typescript source are still appearing quickly

**Pull request merged**